src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSourceContext.java

```diff
@@ -0,0 +1,55 @@
+package com.atlassian.bitbucket.jenkins.internal.scm;
+
+import com.cloudbees.plugins.credentials.Credentials;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import hudson.model.TaskListener;
+import jenkins.scm.api.SCMHeadObserver;
+import jenkins.scm.api.SCMSource;
+import jenkins.scm.api.SCMSourceCriteria;
+import jenkins.scm.api.trait.SCMSourceContext;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static java.util.Objects.requireNonNull;
+
+public class BitbucketSCMSourceContext extends SCMSourceContext<BitbucketSCMSourceContext, BitbucketSCMSourceRequest> {
+
+    private final Credentials credentials;
+    private final Map<BitbucketDiscoverableHeadType, BitbucketSCMHeadDiscoveryHandler> discoveryHandlers =
+            new HashMap<>();
+    private final BitbucketSCMRepository repository;
+
+    private boolean needsChangeRequests;
+
+    public BitbucketSCMSourceContext(SCMSourceCriteria criteria,
+                                     SCMHeadObserver observer,
+                                     @Nullable Credentials credentials,
+                                     BitbucketSCMRepository repository) {
+        super(criteria, observer);
+        this.credentials = credentials;
+        this.repository = repository;
+    }
+
+    public Map<BitbucketDiscoverableHeadType, BitbucketSCMHeadDiscoveryHandler> getDiscoveryHandlers() {
+        return discoveryHandlers;
+    }
+
+    public final boolean needsChangeRequests() {
```

Comment:
I don't think we will need this as we the head discovery would be abstracted by the BitbucketSCMHeadDiscoveryHandler which is added by the relevant discovery trait.